src/patterns.jl

```diff
@@ -0,0 +1,131 @@
+"""
+    AbstractPattern
+
+Abstract supertype of all sparsity pattern representations.
+
+## Type hierarchy
+```
+AbstractPattern
+├── AbstractVectorPattern: used in GradientTracer, ConnectivityTracer
+│   └── IndexSetVectorPattern
+└── AbstractHessianPattern: used in HessianTracer
+    └── IndexSetHessianPattern
+```
+"""
+abstract type AbstractPattern end
+
+#==========================#
+# Utilities on AbstractSet #
+#==========================#
+
+myempty(::Type{S}) where {S<:AbstractSet} = S()
+seed(::Type{S}, i::Integer) where {S<:AbstractSet} = S(i)
+
+""""
+    product(a::S{T}, b::S{T})::S{Tuple{T,T}}
+
+Inner product of set-like inputs `a` and `b`.
+"""
+product(a::AbstractSet{I}, b::AbstractSet{I}) where {I<:Integer} =
+    Set((i, j) for i in a, j in b)
+
+function union_product!(
+    hessian::SH, gradient_x::SG, gradient_y::SG
+) where {I<:Integer,SG<:AbstractSet{I},SH<:AbstractSet{Tuple{I,I}}}
+    hxy = product(gradient_x, gradient_y)
+    return union!(hessian, hxy)
+end
+
+#=======================#
+# AbstractVectorPattern #
+#=======================#
+
+# For use with ConnectivityTracer and GradientTracer.
+
+"""
+    AbstractVectorPattern <: AbstractPattern
+
+Abstract supertype of sparsity patterns representing a vector.
+For use with [`ConnectivityTracer`](@ref) and [`GradientTracer`](@ref).
+
+## Expected interface
+
+* `myempty(::Type{MyPattern})`: return a pattern representing a new number (usually an empty pattern)
+* `seed(::Type{MyPattern}, i::Integer)`: return an pattern that only contains the given index `i`
```

Comment:
These won't work for array-backed patterns because of #137: we need the actual length of the input at some point (perhaps even the size). Input size information should only be stored in the pattern object itself, not in its type (unless we want a compilation nightmare).

Reply:
> Input size information should only be stored in the pattern object itself, not in its type (unless we want a compilation nightmare)

I honestly don't think it would be too bad for realistic use-cases. The input size is constant over the entire sparsity tracing procedure, and therefore only one type gets created.

Reply:
Otherwise, if needed, we can still modify this to `seed(T, i, n)`.

Reply:
That's because you only think of one application of the tracing. What if we need to trace local sparsity on each array of a dataset, each having a different size (e.g. for a dataset of graphs)?

Reply:
Let's do this in a follow up. It's a not a problem introduced by this PR, but one that also exists on main. We'll deal with it before tagging a breaking release.